rfcs/20190117-tf-module.md

```diff
@@ -0,0 +1,634 @@
+express using the `tf.Module` abstraction.
+
+**Layers**
+
+```python
+class Sequential(tf.Module):
+  def __init__(self, layers, name=None):
+    super(Sequential, self).__init__(name=name)
+    self.layers = list(layers)
+
+  def __call__(self, x):
+    for layer in self.layers:
+      x = layer(x)
+    return x
+
+class Linear(tf.Module):
+  def __init__(self, output_size, name=None):
+    super(Linear, self).__init__(name=name)
+    self.output_size = output_size
+
+  def __call__(self, x):
+    x = tf.convert_to_tensor(x, name='x')
+
+    if not hasattr(self, 'w'):
+      input_size = x.shape[1]
+      initial_w = tf.random.truncated_normal(shape=[input_size, self.output_size],
+                                             stddev=(1 / math.sqrt(input_size)))
+      self.w = tf.Variable(initial_w, name='w')
+      self.b = tf.Variable(tf.zeros([self.output_size]), name='b')
+
+    return tf.matmul(x, self.w) + self.b
+
+mlp = Sequential([
+  Linear(1024),
+  tf.nn.relu,
+  Linear(10),
+])
+
+# Calling the module produces the result. In the above example variables
+# are created lazily the first time the module is called.
+y = mlp(tf.ones([10, 100]))
+
+assert set(mlp.variables) == {mlp.layers[0].w, mlp.layers[0].b,
+                              mlp.layers[2].w, mlp.layers[2].b}
+
+# Sequential doesn't own any variables.
+assert not set(mlp.owned_variables)
+
+# Variable names include module names.
+assert mlp.layers[0].w.name == 'linear/w:0'
+assert mlp.layers[2].w.name == 'linear/w:0'  # NOTE: in graph mode this is linear_1
+
+# Sequential depends directly on it's layers so these are it's children.
+assert set(mlp.owned_submodules) == {mlp.layers[0], mlp.layers[2]}
+
+# The full submodule graph only includes the two linear layers.
+assert set(mlp.submodules) == {mlp.layers[0], mlp.layers[2]}
+```
+
+**Metrics**
+
+```python
+class Mean(tf.Module):
+  def __init__(self, dtype=tf.float32, name=None):
+    super(Mean, self).__init__(name=name)
+    self.count = tf.Variable(0., dtype=dtype, name='count')
+    self.sum = tf.Variable(0., dtype=dtype, name='sum')
+
+  def reset(self):
+    self.count.assign(0)
+    self.sum.assign(0)
+
+  def update(self, sample):
+    self.sum.assign_add(sample)
+    self.count.assign_add(1)
+
+  def result(self):
+    return self.sum / self.count
+```
+
+**Optimizers**
+
+```python
+class GradientDescent(tf.Module):
+  def __init__(self, learning_rate=0.01, name=None):
+    super(GradientDescent, self).__init__(name=name)
+    self.learning_rate = tf.Variable(learning_rate, name='learning_rate')
+
+  def minimize(self, loss_fn, variables=None):
+    with tf.GradientTape() as tape:
+      loss = loss_fn()
+    if variables is None:
+      variables = tape.watched_variables()
+    self.apply_gradients(zip(variables, tape.gradients(loss, variables)))
+    return loss
+
+  def apply_gradients(self, vars_and_grads):
+    for variable, grad in vars_and_grads:
+      variable.assign_sub(grad * self.learning_rate)
+```
```

Comment:
I'd like to comment here on the following points:

 1. Module should be the base class for tf.keras.Layer
 2. We explicitly chose not to force a call signature or build() strategy
 3. Pros and cons of the name Module (familiarity with pytorch being a big issue here)

Reply:
Agreed re (1) and re (2) I call out why a bit in the detailed design if anyone wants more context.

Re (3) I feel strongly that `Module` is the right name and what users will expect. The term `Module` and its usage in ML is almost as old as I am, the oldest reference I know about (h/t @jekbradbury for the reference) is from SN/Lush [0]:

> In Lush, building and training a complex system is done by assembling basic blocks, called modules. Modules are subclasses of the class gb-module.

[0] http://lush.sourceforge.net/lush-manual/51d13a16626fd68f.html